File: lena-config/src/main/java/org/riversoforion/lena/config/Namespace.java
/*
 * Copyright (c) 2024-2025. Eric McIntyre / Rivers of Orion
 */
package org.riversoforion.lena.config;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import static java.util.Arrays.asList;
import static java.util.Arrays.stream;

/**
 * Configuration namespace, including property names. A normalized namespace is as follows:
 * <ul>
 *     <li></li>
 * </ul>
 */
// TODO Document
public class Namespace implements Comparable<Namespace>, CharSequence {

    private final Name name;
    private final String fullName;

    private Namespace(List<String> segments) {

        this.name = Name.forNamespace(segments);
        this.fullName = Name.SEPARATOR + name;
    }

    // Static factories

    public static Namespace root() {

        return of();
    }

    public static Namespace of(String... segments) {

        if (segments == null) {
            segments = new String[0];
        }
        return new Namespace(stream(segments).filter(Namespace::notEmpty).flatMap(Name::splitSegment).toList());
    }

    public static Namespace parse(String name) {

        String[] segments = name.split(Name.SEPARATOR);
        if (segments.length == 1 && segments[0].isEmpty()) {
            return root();
        }
        return new Namespace(asList(segments));
    }

    private static boolean notEmpty(String segment) {

        return segment != null && !segment.isBlank();
    }

    // Operations

    public String name() {

        return fullName;
    }

    public List<String> segments() {

        return name.segments();
    }

    public Namespace child(String name) {

        if (name == null || name.isBlank()) {
            throw new IllegalArgumentException("Child namespace must contain at least one new segment");
        }
        if (name.startsWith(Name.SEPARATOR)) {
            return Namespace.parse(name);
        }
        List<String> childSegments = asList(name.split(Name.SEPARATOR));
        if (childSegments.isEmpty()) {
            throw new IllegalArgumentException("Child namespace must contain at least one new segment");
        }
        List<String> allSegments = new ArrayList<>(this.name.segments());
        allSegments.addAll(childSegments);
        return new Namespace(allSegments);
    }

    public List<String> resolveProperty(Name propName) {

        if (propName == null) {
            throw new IllegalArgumentException("Property name required");
        }
        List<String> property = new ArrayList<>(this.name.segments());
        property.addAll(propName.segments());
        return property;
    }

    // Object

    @Override
    public String toString() {

        return name();
    }

    @Override
    public boolean equals(Object o) {

        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        Namespace namespace = (Namespace) o;
        return Objects.equals(fullName, namespace.fullName);
    }

    @Override
    public int hashCode() {

        return Objects.hashCode(fullName);
    }

    // CharSequence

    @Override
    public int length() {

        return fullName.length();
    }

    @Override
    public char charAt(int index) {

        return fullName.charAt(index);
    }

    @Override
    public CharSequence subSequence(int start, int end) {

        return fullName.subSequence(start, end);
    }

    // Comparable

    @Override
    public int compareTo(Namespace o) {

        if (o == null) {
            return 1;
        }
        return fullName.compareTo(o.fullName);
    }
}

File: lena-config/src/commonMain/kotlin/org/riversoforion/lena/config/Namespace.kt
/*
 * Copyright (c) 2024-2026. Eric McIntyre / Rivers of Orion
 */
package org.riversoforion.lena.config

/**
 * A normalized configuration namespace (scoping prefix).
 *
 * Always begins with [Name.SEPARATOR]. The root namespace is represented as `"/"`.
 * Segments follow the same normalization rules as [Name].
 *
 * Examples:
 * - `Namespace.root()` → `"/"`
 * - `Namespace.of("app", "service")` → `"/app/service"`
 * - `Namespace.of("app").child("network")` → `"/app/network"`
 */
public class Namespace private constructor(
    private val name: Name,
    private val fullName: String,
) : Comparable<Namespace> {

    public companion object {
        public fun root(): Namespace = of()

        public fun of(vararg segments: String): Namespace {
            val flat = segments.filter { it.isNotBlank() }.flatMap { Name.splitSegment(it) }
            return fromSegmentList(flat)
        }

        public fun parse(name: String): Namespace {
            val segments = name.split(Name.SEPARATOR)
            return if (segments.size == 1 && segments[0].isEmpty()) root()
            else fromSegmentList(segments)
        }

        private fun fromSegmentList(segments: List<String>): Namespace {
            val name = Name.forNamespace(segments)
            val fullName = Name.SEPARATOR + name.toString()
            return Namespace(name, fullName)
        }
    }

    public fun name(): String = fullName

    public fun segments(): List<String> = name.segments()

    public fun child(childName: String): Namespace {
        require(childName.isNotBlank()) { "Child namespace must contain at least one new segment" }
        return if (childName.startsWith(Name.SEPARATOR)) {
            parse(childName)
        } else {
            val childSegments = childName.split(Name.SEPARATOR).filter { it.isNotBlank() }
            require(childSegments.isNotEmpty()) { "Child namespace must contain at least one new segment" }
            fromSegmentList(name.segments() + childSegments)
        }
    }

    public fun resolveProperty(propName: Name): List<String> {
        return name.segments() + propName.segments()
    }

    override fun toString(): String = fullName

    override fun equals(other: Any?): Boolean {
        if (other !is Namespace) return false
        return fullName == other.fullName
    }

    override fun hashCode(): Int = fullName.hashCode()

    override fun compareTo(other: Namespace): Int = fullName.compareTo(other.fullName)
}
